### ISSUE TITLE
JSON API for classes

### ISSUE THREAD

**Issue by JasonGross** [learning-unlimited/ESP-Website#3152]
I like adding my classes to Google Calendar, but this is quite painful when I have to scrape the html manually and minor formatting details on the website change every year, and some years some pages (such as class descriptions) give 404s.

#3151 would help when I only have a few classes, but I'd also like to be able to access a JSON API for reading the details of my classes, e.g., https://esp.mit.edu/teach/Splash/2020/teacherreg/json would give me a JSON page containing something like:
{"program": "Splash",
"year": 2020,
"month": "November",
"firstDay": 14,
"classes": [{"name": "Proving Löb's Theorem", "number": 14173, category: "M", "description": ..., "teachers": ["Jason Gross", "Allison Strandberg"], "status": "Accepted", "number of students": 15, "time blocks": [{"day": 15, "startTime": "11:05am", "endTime": "11:55am", "location": "Zoom Room 14", "sudent_list_url": "https://esp.mit.edu/teach/Splash/2020/section_students/20222"}, ...]}, ... ]
}

I'm fine with pretty much whatever format, as long as it contains the relevant information, as I just want to parse it so that I can add the classes to my google calendar.

Is this possible?

I'm also happy to contribute to developing this if you can point me at the right files to add/edit/modify, and some examples of how to access the relevant data.

**Comment by willgearty** [comment]
This should be pretty simple. We currently have a page (`/tl/one/two/teacherschedule`) that returns an HTML formatted teacher schedule (view [here](https://github.com/learning-unlimited/ESP-Website/blob/417c66cb7b0c6148a60b4f26851debb2d1f5593c/esp/esp/program/modules/handlers/teacherpreviewmodule.py#L84), which calls the method [here](https://github.com/learning-unlimited/ESP-Website/blob/417c66cb7b0c6148a60b4f26851debb2d1f5593c/esp/esp/program/modules/handlers/teacherpreviewmodule.py#L59)). If someone specified `/tl/one/two/teacherschedule/json` (in this case, "json" would be the `extra` argument for the method), we could catch that and just return some json about the schedule instead of a rendered page.

I don't imagine it would get used much, so really not a high priority, but you're welcome to submit a pull request. Let me know if you need more direction!

**Comment by JasonGross** [comment]
Thanks!  I'll look into that.  Which `teacherschedule.html` is the template used?  And is this page accessible to teachers?  (https://esp.mit.edu/teach/Splash/2020/teacherschedule is a 404, and I don't see a likely-looking link on https://esp.mit.edu/teach/Splash/2020/teacherreg)

**Comment by willgearty** [comment]
The default schedule uses the template [here](https://github.com/learning-unlimited/ESP-Website/blob/main/esp/templates/program/modules/programprintables/teacherschedule.html). I'm guessing MIT doesn't have the module enabled ("Teacher preview module") so the link doesn't work.

**Comment by SuyashJain17** [comment]
If this issue is still open and needs work, I’d be happy to take it on. Could you please assign it to me? @JasonGross @willgearty

**Comment by willgearty** [comment]
@SuyashJain17 this particular use case may be covered by #3151. I'm not sure we want to be supporting a public-facing API at this time.

**Comment by Supriyo760** [comment]
Hey @willgearty 
a public-facing API isn't currently desired and that #3151 covers the primary calendar use case, should we go ahead and close this issue? 
If the team decides to explore a more formalized REST API approach in the future, we can consolidate any further discussion on JSON endpoints into the newly opened #4689 instead. Let me know if that makes sense!

**Comment by nishantharkut** [comment]
Agreed, @Supriyo760. Since #3151 covers the primary calendar use case mentioned by the author, closing this helps keep the backlog clean. I've opened #4689 to specifically address the module-management endpoints needed for the new UI project (#3854), so we can consolidate the REST API discussion there. Thanks!
